### PR TITLE
[WIP] Create initial config definition

### LINKED ISSUE #143: Strictly validate autorc config before any release
**Is your feature request related to a problem? Please describe.**

A few of us at Artsy are working on migrating a lot of repos over to use auto-release. In a batch session, I somehow managed to mess up the config and used `noReleaseLabels` instead of `skipReleaseLabels` on several of the projects. This caused some unnecessary releases. 

**Describe the solution you'd like**

I think the config should be strictly validated. If there are values present that aren't a part of the config (i.e. noReleaseLabels), the release process should fail. Invalid configuration likely means someone is in for a bad time. Similarly if a config object doesn't match an expected type, etc. It seems like there's a start to a json schema that could be used for this purpose ([auto-rc.json](https://github.com/intuit/auto-release/blob/master/src/auto-rc.json)). 

**Describe alternatives you've considered**

If you want to take a lighter approach I'd recommend at the very least having a `validate` command that checks the validity of the configuration.

### PR DESCRIPTION
**Pull request by zephraph**
Fixes #143

# What Changed

* Adds `config.ts` that acts as a schema for the autorc configuration
* Adds a build process to generate a json schema from `config.ts`

# Why

This started as a resolution to #143. If there's a typo or invalid property added to an `.autorc` we want to alert people and fail before letting their release process run amok and do god only knows what. This could also be a way to auto generate documentation for the autorc (the docs are kinda wrong currently for some things and woefully incomplete) and could assist in building a vscode plugin. The possibilities are only limited by my laziness. 

Todo:

- [ ] Add tests
- [ ] Add docs
- [x] Add yourself to contributors (run `yarn contributors:add`)

### CODE REVIEW

**Comment by zephraph** [comment]
The user now gets a pretty useful error when an invalid property is present. 

image

**Comment by hipstersmoothie** [comment]
That emoji is great lol

**Comment by hipstersmoothie** [comment]
@zephraph plugin config definition changed a little https://github.com/intuit/auto-release/blob/86173e751212985362704cbd4f34cd9788d3ba5c/src/github-release.ts#L39

**Comment by zephraph** [comment]
@hipstersmoothie when you get a minute, will you peek at the `config.ts`? I'm thinking ultimately it may replace `IGithubReleaseOptions`, `VersionLabel`, all the defaults, etc. A peek at the general direction would be good. I've got more changes I can push when I get home that moves in this direction.

**Comment by hipstersmoothie** [comment]
I love this! replacing all that code would be awesome!

**Comment by zephraph** [comment]
I've got a lot of type and config updates left to make. I'm just pushing up what I have now so I can work on it tomorrow on my other machine. Trying to wrap it up for the 3.0 release.

File: src/config.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+export class Labels {
+  public [Label.major]: string = 'major';
+  public [Label.minor]: string = 'minor';
+  public [Label.patch]: string = 'patch';
+  public [Label.skipRelease]: string = 'skip-release';
+  /**
+   * @deprecated
+   */
+  public 'skip-release': string = 'skip-release';
```
Comment (hipstersmoothie) [inline src/config.ts]: That makes sense 👍

**Comment by hipstersmoothie** [comment]
Interesting way to use classes, I wouldn't have though of that

File: src/main.ts [hunk 2]
```diff
@@ -189,19 +191,10 @@ export class AutoRelease {
...
-          [
-            ...Object.keys(new ChangelogTitles()),
-            ...Object.keys(
-              this.githubRelease.releaseOptions.changelogTitles || {}
-            )
-          ].map((label): [string, string] => [label, label])
-        )
-      ])
-    );
+    await this.githubRelease.addLabelsToProject({
```
Comment (hipstersmoothie) [inline src/main.ts]: 😍

**Comment by zephraph** [comment]
> Interesting way to use classes, I wouldn't have thought of that

Interesting isn't always good, heh. The reason why I did it this way initially was because `quicktype` will infer the initial values of the members of the class as default values for the json schema. I'm not convinced it's the right way to approach this.

**Comment by zephraph** [comment]
It's still a little slow going. There's a few things...

1. I feel like how the config is set up currently isn't great. There's a few different representations of config in the actual program. What comes from the user, the default values, and what gets passed to the release process (with cli agruments and other stuff added in). Somehow I'd like to do a better job of separating all these things out. As a part of this, decoupling configuration from any specific provider (i.e. github) will make gitlab integration easier later.  I'll poke on it some more tomorrow.

**Comment by zephraph** [comment]
Being as this PR is pretty big and a lot of changes happening for v3 I'm just going to close this. We can incrementally add some of these back.

**Pull request closed without merging**